### PR TITLE
HIVE-26704: Cleaner shouldn't be blocked by global min open txnId

### PR DESCRIPTION
**Pull request by deniskuzZ**
### What changes were proposed in this pull request?

Removes a bottleneck in the Compaction process

### Why are the changes needed?

Currently, if there is a single long-running transaction that can prevent the Cleaner to clean up any tables. This causes file buildup in tables, which can cause performance penalties when listing the directories 

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Unit tests

### CODE REVIEW

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/CompactionTxnHandler.java [hunk 1]
```diff
@@ -384,29 +384,47 @@ public List<CompactionInfo> findReadyToClean(long minOpenTxnWaterMark, long rete
          * By filtering on minOpenTxnWaterMark, we will only cleanup after every transaction is committed, that could see
          * the uncompacted deltas. This way the cleaner can clean up everything that was made obsolete by this compaction.
          */
-        String whereClause = " WHERE \"CQ_STATE\" = '" + READY_FOR_CLEANING + "'";
-        if (minOpenTxnWaterMark > 0) {
+        String whereClause = " WHERE \"CQ_STATE\" = " + quoteChar(READY_FOR_CLEANING) +
```
Comment (rkirtir) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/CompactionTxnHandler.java:387]: Is it possible to move this bigger query to TxnQueries Class and use .append()

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/CompactionTxnHandler.java [hunk 2]
```diff
@@ -1577,24 +1590,20 @@ public long findMinTxnIdSeenOpen() throws MetaException {
             }
             minOpenTxn = rs.getLong(1);
             if (rs.wasNull()) {
-              minOpenTxn = -1L;
+              minOpenTxn = Long.MAX_VALUE;
```
Comment (SourabhBadhya) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/CompactionTxnHandler.java:1593]: Can we use ternary operator here - 
`minOpenTxn = rs.wasNull() ? Long.MAX_VALUE : rs.getLong(1)`

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HMSHandler.java [hunk 3]
```diff
@@ -76,6 +76,7 @@
 import java.lang.reflect.UndeclaredThrowableException;
 import java.nio.ByteBuffer;
 import java.security.PrivilegedExceptionAction;
+import java.sql.SQLException;
```
Comment (SourabhBadhya) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HMSHandler.java:79]: nit: Unused import

File: ql/src/java/org/apache/hadoop/hive/ql/lockmgr/DbTxnManager.java [hunk 4]
```diff
@@ -393,6 +397,32 @@ private boolean allowOperationInATransaction(QueryPlan queryPlan) {
...
 
+  @Override
+  public void addWriteIdsToMinHistory(QueryPlan plan, ValidTxnWriteIdList txnWriteIds) {
+    if (plan.getInputs().isEmpty()) {
+      return;
+    }
+    Map<String, Long> writeIds = plan.getInputs().stream()
+      .filter(input -> !input.isDummy() && AcidUtils.isTransactionalTable(input.getTable()))
+      .map(input -> input.getTable().getFullyQualifiedName())
+      .collect(Collectors.toSet()).stream()
```
[the remark this answers is not in the document]
Comment (deniskuzZ) [reply ql/src/java/org/apache/hadoop/hive/ql/lockmgr/DbTxnManager.java:408]: so we don't call an expensive getMinOpenWriteId multiple times for the same table

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/CompactionTxnHandler.java (hunk 2 above)
Comment (deniskuzZ) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/CompactionTxnHandler.java:1593]: nope, per javaDoc wasNull reports whether the last column read had a value of SQL <code>NULL</code>. You must first call one of the getter methods on a column to try to read its value and then call the method <code>wasNull</code> to see if the value read was SQL <code>NULL</code>.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HMSHandler.java (hunk 3 above)
Comment (deniskuzZ) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HMSHandler.java]: fixed

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/CompactionTxnHandler.java (hunk 2 above)
Comment (SourabhBadhya) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/CompactionTxnHandler.java:1603]: Ok

File: ql/src/java/org/apache/hadoop/hive/ql/lockmgr/DbTxnManager.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (SourabhBadhya) [reply ql/src/java/org/apache/hadoop/hive/ql/lockmgr/DbTxnManager.java:410]: Makes sense.

File: ql/src/test/org/apache/hadoop/hive/ql/txn/compactor/CompactorTest.java [hunk 5]
```diff
@@ -133,6 +133,7 @@ protected final void setup(HiveConf conf) throws Exception {
     MetastoreConf.setTimeVar(conf, MetastoreConf.ConfVars.TXN_OPENTXN_TIMEOUT, 2, TimeUnit.SECONDS);
     MetastoreConf.setBoolVar(conf, MetastoreConf.ConfVars.COMPACTOR_INITIATOR_ON, true);
     MetastoreConf.setBoolVar(conf, MetastoreConf.ConfVars.COMPACTOR_CLEANER_ON, true);
+    MetastoreConf.setBoolVar(conf, MetastoreConf.ConfVars.TXN_USE_MIN_HISTORY_WRITE_ID, false);
```
Comment (SourabhBadhya) [inline ql/src/test/org/apache/hadoop/hive/ql/txn/compactor/CompactorTest.java:136]: If we set it as false, is there any test which run using the new logic?
CompactorTest is extended by a lot of other Test classes (TestCleaner, TestCleaner2, TestCleanerWithoutMinHistoryLevel etc).

Comment (deniskuzZ) [reply ql/src/test/org/apache/hadoop/hive/ql/txn/compactor/CompactorTest.java]: added new test

File: ql/src/java/org/apache/hadoop/hive/ql/lockmgr/DbTxnManager.java [hunk 6]
```diff
@@ -395,6 +399,32 @@ private boolean allowOperationInATransaction(QueryPlan queryPlan) {
...
```
Comment (veghlaci05) [inline ql/src/java/org/apache/hadoop/hive/ql/lockmgr/DbTxnManager.java:410]: If I'm right, and this is to eliminate duplicates, than you could use the distinct() method.

File: ql/src/java/org/apache/hadoop/hive/ql/lockmgr/DbTxnManager.java [hunk 7]
```diff
@@ -395,6 +399,32 @@ private boolean allowOperationInATransaction(QueryPlan queryPlan) {
...
+    if (!writeIds.isEmpty()) {
+      try {
+        getMS().addWriteIdsToMinHistory(txnId, writeIds);
+      } catch (TException | LockException e) {
+        throw new RuntimeException(ErrorMsg.METASTORE_COMMUNICATION_FAILED.getMsg(), e);
+      }
+    }
+  }
+
+  private Long getMinOpenWriteId(ValidTxnWriteIdList txnWriteIds, String table) {
```
Comment (veghlaci05) [inline ql/src/java/org/apache/hadoop/hive/ql/lockmgr/DbTxnManager.java:422]: The only usage of this method is above, in situations like this I prefer to embed it.

File: ql/src/test/org/apache/hadoop/hive/ql/txn/compactor/TestCleaner2.java [hunk 8]
```diff
@@ -23,6 +23,6 @@
 public class TestCleaner2 extends TestCleaner {
   @Override
   boolean useHive130DeltaDirName() {
-    return false;
+    return true;
```
Comment (veghlaci05) [inline ql/src/test/org/apache/hadoop/hive/ql/txn/compactor/TestCleaner2.java:26]: Why is this change? Is this related to the PR?

**Review by veghlaci05 — CHANGES_REQUESTED** [review]

File: ql/src/test/org/apache/hadoop/hive/ql/txn/compactor/TestCleaner2.java (hunk 8 above)
Comment (deniskuzZ) [reply ql/src/test/org/apache/hadoop/hive/ql/txn/compactor/TestCleaner2.java:26]: no, but we have TestCleaner & TestCleaner2 running the same stuff

File: ql/src/java/org/apache/hadoop/hive/ql/lockmgr/DbTxnManager.java (hunk 7 above)
Comment (deniskuzZ) [reply ql/src/java/org/apache/hadoop/hive/ql/lockmgr/DbTxnManager.java:422]: it's used in a lambda

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java [hunk 9]
```diff
@@ -5894,6 +5902,63 @@ private void addTxnToMinHistoryLevel(Connection dbConn, List<Long> txnIds, long
...
+                "Batch size: " + maxBatchSize);
+              pstmt.executeBatch();
+            }
+          }
+          if (writeId % maxBatchSize != 0) {
+            LOG.debug("Executing a batch of <" + TXN_TO_WRITE_ID_INSERT_QUERY + "> queries. " +
+              "Batch size: " + writeId % maxBatchSize);
+            pstmt.executeBatch();
+          }
+        }
```
[the remark this answers is not in the document]
Comment (deniskuzZ) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java:5940]: that would be a massive refactor of TxnHandler

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Cleaner.java [hunk 10]
```diff
@@ -140,41 +140,36 @@ public void run() {
...
               CompletableFuture<Void> asyncJob =
                   CompletableFuture.runAsync(
-                          ThrowingRunnable.unchecked(() -> clean(compactionInfo, cleanerWaterMark, metricsEnabled)),
-                          cleanerExecutor)
-                      .exceptionally(t -> {
-                        LOG.error("Error clearing {}", compactionInfo.getFullPartitionName(), t);
-                        return null;
-                      });
+                      ThrowingRunnable.unchecked(() -> {
+                        long minOpenTxnGLB = (ci.minOpenWriteId > 0) ?
```
[the remark this answers is not in the document]
Comment (deniskuzZ) [reply ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Cleaner.java]: fixed

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (veghlaci05) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java:5940]: Ok, I'll create a separate ticket for it. It mag get picked up eventually

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/CompactionTxnHandler.java [hunk 11]
```diff
@@ -1535,20 +1556,14 @@ public void setHadoopJobId(String hadoopJobId, long id) {
...
-        rollbackDBConn(dbConn);
-        checkRetryable(e, "getMinOpenTxnForCleaner");
-        throw new MetaException("Unable to execute getMinOpenTxnIfForCleaner() " +
-            e.getMessage());
-      } finally {
-        closeDbConn(dbConn);
-      }
+        LOG.error("Unable to findMinOpenTxnIdForCleaner", e);
+        checkRetryable(e, "findMinOpenTxnIdForCleaner");
+        throw new MetaException("Unable to execute getMinOpenTxnIfForCleaner() " + e.getMessage());
```
Comment (SourabhBadhya) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/CompactionTxnHandler.java:1565]: nit: Typo - `getMinOpenTxnIdForCleaner` instead of `getMinOpenTxnIfForCleaner`

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Cleaner.java [hunk 12]
```diff
@@ -140,41 +140,36 @@ public void run() {
...
                   CompletableFuture.runAsync(
-                          ThrowingRunnable.unchecked(() -> clean(compactionInfo, cleanerWaterMark, metricsEnabled)),
-                          cleanerExecutor)
-                      .exceptionally(t -> {
-                        LOG.error("Error clearing {}", compactionInfo.getFullPartitionName(), t);
-                        return null;
-                      });
+                      ThrowingRunnable.unchecked(() -> {
+                        long minOpenTxn = (ci.minOpenWriteId > 0) ? 
+                            ci.nextTxnId + 1 : Math.min(minOpenTxnId, txnHandler.findMinTxnIdSeenOpen());
```
[the remark this answers is not in the document]
Comment (deniskuzZ) [reply ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Cleaner.java:165]: yea, missed that, if we have minOpenWriteId, wee should call findMinTxnIdSeenOpen

[the remark this answers is not in the document]
Comment (deniskuzZ) [reply ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Cleaner.java]: fixed

**Review by veghlaci05 — APPROVED** [review]
LGTM approved, pending tests.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/CompactionTxnHandler.java (hunk 11 above)
Comment (deniskuzZ) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/CompactionTxnHandler.java]: fixed
